### ISSUE TITLE
Plugin security is abysmal... thoughts on possible solution

### ISSUE THREAD

**Issue by molecular** [Electron-Cash/Electron-Cash#2262]
I'm developing a plugin and the security model is, well... non-existant.

Users need to trust plugins. They can do anything.

Apart from fixing that (probably prohibitively cumbersome if possible at all), maybe a vouching scheme could be implemented in EC roughly as follows:

Hardcode keys of trusted community members (devs). Those guys could sign messages like 

  * "I looked at plugin.zip with hash <xzy> and I couldn't find any obvious malicious code" or even
  * "I did a full code audit of plugin.zip and no malicious behaviour was found" or even
  * "I've been using this plugin for a while and my funds are still there"

or of course

  * "This plugin has malicous code. Will steal your funds and erase your HD"
 
Those messages would have to be distributed somehow (either in parallel to the plugin.zip, or through some centralized service or maybe even through the blockchain?).

Then the plugin installation dialog could display them when the user tries to install an external plugin.

Is that viable? Good idea?

**Comment by cculianu** [comment]
Yeah you're right. I think that would be a step towards trust.  Now the question is -- is it enforced by the software or just "recommended" and you can "continue anyway" -- or is it completely external to the software (like you just check a webpage).

Another alternative would be to just have a website with signed trusted plugins that we vet. (Or maybe not an alternative just an additional thing).

You are right in that -- it is a lot of trust to install one and right now there are no steps taken, not even minimal ones.

**Comment by molecular** [comment]
thanks for chiming in, @cculianu 

> is it enforced by the software or just "recommended" and you can "continue anyway"

I wouldn't enforce it. There might be plugins that didn't get reviewd by anyone yet. It shouldn't be mandatory to get a review by the "select few" that can do it.

> or is it completely external to the software (like you just check a webpage).

well, my idea is that there are hardcoded keys of reviewers (or whatever we call them) in EC. This is like for example archlinux signs binaries: there's a (changing) list of keys that have signing power. This builds on the fact that users are *already* trusting the OS (or EC in this case). Of course "just a webpage" might work, but that page still needs to earn the trust EC already has.

> Another alternative would be to just have a website with signed trusted plugins that we vet.

who is "we", though? How does the user of that webpage know it's by the "EC devs"?

> You are right in that -- it is a lot of trust to install one and right now there are no steps taken, not even minimal ones.

The current situation sucks. It would be very easy to copy an existing popular plugin (put it on githubalternative.com). No amount of key checking will help the user against this attack. But a "this is the popular inter-wallet-transfer plugin everyone uses, signed calin (calin key is hardcoded in EC).

Let me say that the whole idea does not really "solve the problem". It also creates centralized authority of sorts (that's why I'm not sure I like the idea of an "EC App Store"). But it's better than what people do now: just *hope* the plugin isn't malicious and *maybe* check the hashes (against what? where do they get the hashes?)

The very least we could do is somehow publish lists of hashes of malicious plugins and warn from the plugin installer if a plugin is on the list, then include an easy reporting function that creates an issue for us with the plugin included so it can be checked.

Oh and I've been thinking: the whole thing could be done as a plugin ;-)
